### PR TITLE
Storage: Refactor DMContext to avoid being misused

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/8312

Problem Summary:

### What is changed and how it works?

* Rename `DMContext::db_context` to `DMContext::global_context`
* Always save global_context as member in DMContext
* Remove a lock in `DeltaMergeStore::newDMContext` that does not have concurrent issue in the latest code

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp [hunk 1]
```diff
@@ -449,13 +449,11 @@ DMContextPtr DeltaMergeStore::newDMContext(
     const String & tracing_id,
     ScanContextPtr scan_context_)
 {
-    std::shared_lock lock(read_write_mutex);
-
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp:453]: The lock was added to protect the "schema" in very old versions(https://github.com/pingcap/tiflash/pull/507/files#diff-22b900e9e8020dc835612316f3bf151cae28b621bfac64d6a22b377d062f4b7eR165-R190). But now the members used in this function do not 
have concurrent issues. So this lock can be removed.

File: dbms/src/Storages/DeltaMerge/DMContext.h [hunk 2]
```diff
@@ -103,9 +162,9 @@ struct DMContext : private boost::noncopyable
         bool is_common_handle_,
         size_t rowkey_column_size_,
         const DB::Settings & settings,
-        const ScanContextPtr scan_context_ = nullptr,
+        const ScanContextPtr & scan_context_ = nullptr,
         const String & tracing_id_ = "")
-        : db_context(db_context_)
+        : global_context(session_context_.getGlobalContext()) // always save the global context
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/DMContext.h:167]: always save the global context no matter what the pass-in `session_context` is a session only context or global context, and rename it to be `global_context`

**Review by Lloyd-Pottiger — APPROVED** [review]

**Pull request merged**